pyomo/contrib/gdpopt/gloa.py

```diff
@@ -0,0 +1,225 @@
+        Disjunctive Programming and Applications to Process Systems.
+        Comp. and Chem. Eng. 2001, 25, 1675-1697.
+        DOI: 10.1016/S0098-1354(01)00732-3.
+        """.strip())
+
+    def solve(self, model, **kwds):
+        """Solve the model.
+
+        Args:
+            model (Block): a Pyomo model or block to be solved
+
+        """
+        return super().solve(model, **kwds)
+
+    def _solve_gdp(self, original_model, config):
+        logger = config.logger
+
+        # we need to gather a map of Disjuncts to their active Constraints
+        # before we call any GDP transformations, as we will need this
+        # information for cut generation later
+        add_constraints_by_disjunct(self.original_util_block)
+        # We also save these in advance because we know only linear logical
+        # constraints will be added by the transformation to a MIP, so these are
+        # all we'll ever need.
+        add_global_constraint_list(self.original_util_block)
+        (master_util_block,
+         subproblem_util_block) = _get_master_and_subproblem(self, config)
+
+        master = master_util_block.model()
+        subproblem = subproblem_util_block.model()
+        master_obj = next(master.component_data_objects(Objective, active=True,
+                                                        descend_into=True))
+
+        self._log_header(logger)
+
+        # main loop
+        while self.iteration < config.iterlim:
+            self.iteration += 1
+
+            # solve linear master problem
+            with time_code(self.timing, 'mip'):
+                mip_feasible = solve_MILP_master_problem(master_util_block,
+                                                         self, config)
+                self._update_bounds_after_master_problem_solve(
+                    mip_feasible, master_obj, logger)
+            # Check termination conditions
+            if self.any_termination_criterion_met(config):
+                break
+
+            with time_code(self.timing, 'nlp'):
+                self._fix_master_soln_solve_subproblem_and_add_cuts(
+                    master_util_block, subproblem_util_block, config)
+
+            # Add integer cut
+            with time_code(self.timing, "integer cut generation"):
+                add_no_good_cut(master_util_block, config)
+
+            # Check termination conditions
+            if self.any_termination_criterion_met(config):
+                break
+
+    # Utility used in cut_generation: We saved a map of Disjuncts to the
+    # active constraints they contain on the master problem util_block, and use
+    # it here to find the active constraints under the current discrete
+    # solution. Note that this preprocesses not just to be efficient, but
+    # because everything on the Disjuncts is deactivated at this point, since
+    # we've already transformed the master problem to a MILP
+    def _get_active_untransformed_constraints(self, util_block, config):
```

Comment:
Is this utility only used here? Maybe moving this to utils.py might make sense as it can be useful for LOA as well.

Reply:
I think that this can wait for a future PR, maybe even consider having the same util for MindtPy. There is going to be a MindtPy rewrite coming after this is merged where we can see where things make sense together.